src/SDKs/Azure.Base/data-plane/Azure.Base.Tests/FixedRetryPolicyTests.cs

```diff
@@ -0,0 +1,228 @@
+            var task = SendRequest(mockTransport, policy);
+
+            await mockTransport.RequestGate.Cycle(new MockResponse(500));
+
+            var delay = await policy.DelayGate.Cycle();
+            Assert.AreEqual(delay, TimeSpan.FromSeconds(3));
+
+            await mockTransport.RequestGate.Cycle(new MockResponse(501));
+
+            var response = await task.TimeoutAfterDefault();
+            Assert.AreEqual(501, response.Status);
+        }
+
+        [Test]
+        public async Task RetriesOnException()
+        {
+            var policy = new FixedRetryPolicyMock(retriableCodes: new [] { 500 }, exceptionFilter: ex => ex is InvalidOperationException, delay: TimeSpan.FromSeconds(3));
+            var mockTransport = new MockTransport();
+            var task = SendRequest(mockTransport, policy);
+
+            await mockTransport.RequestGate.CycleWithException(new InvalidOperationException());
+
+            var delay = await policy.DelayGate.Cycle();
+            Assert.AreEqual(delay, TimeSpan.FromSeconds(3));
+
+            await mockTransport.RequestGate.Cycle(new MockResponse(200));
+
+            var response = await task.TimeoutAfterDefault();
+            Assert.AreEqual(200, response.Status);
+        }
+
+        [Test]
+        public async Task RetriesOnlyFilteredException()
+        {
+            var policy = new FixedRetryPolicyMock(retriableCodes: new [] { 500 }, exceptionFilter: ex => ex is InvalidOperationException, delay: TimeSpan.FromSeconds(3));
+            var mockTransport = new MockTransport();
+            var task = SendRequest(mockTransport, policy);
+
+            await mockTransport.RequestGate.CycleWithException(new InvalidOperationException());
+
+            var delay = await policy.DelayGate.Cycle();
+            Assert.AreEqual(delay, TimeSpan.FromSeconds(3));
+
+            await mockTransport.RequestGate.CycleWithException(new IOException());
+
+            var exception = Assert.ThrowsAsync<AggregateException>(async () => await task.TimeoutAfterDefault());
+            StringAssert.StartsWith("Retry failed after 2 tries.", exception.Message);
+            Assert.IsInstanceOf<InvalidOperationException>(exception.InnerExceptions[0]);
+            Assert.IsInstanceOf<IOException>(exception.InnerExceptions[1]);
+        }
+
+        [Test]
+        public async Task RetriesOnlyFilteredExceptionFirst()
+        {
+            var policy = new FixedRetryPolicyMock(retriableCodes: new [] { 500 }, exceptionFilter: ex => ex is InvalidOperationException, delay: TimeSpan.FromSeconds(3));
+            var mockTransport = new MockTransport();
+            var task = SendRequest(mockTransport, policy);
+
+            await mockTransport.RequestGate.CycleWithException(new IOException());
+
+            var exception = Assert.ThrowsAsync<AggregateException>(async () => await task.TimeoutAfterDefault());
+            StringAssert.StartsWith("Retry failed after 1 tries.", exception.Message);
```

Comment:
[...] after 1 try. ?

Reply:
I'm not sure that we want to wrap this exception at all.

Reply:
Because this means that all exceptions coming through the pipeline would be wrapped by AggregateException which is not the greatest experience.